dbms/src/Functions/greatCircleDistance.cpp

```diff
@@ -14,70 +14,80 @@
 namespace DB
 {
 
-namespace ErrorCodes
-{
-    extern const int ARGUMENT_OUT_OF_BOUND;
-    extern const int ILLEGAL_COLUMN;
-    extern const int LOGICAL_ERROR;
-}
-
-/** https://en.wikipedia.org/wiki/Great-circle_distance
- *
- *  The function calculates distance in meters between two points on Earth specified by longitude and latitude in degrees.
- *  The function uses great circle distance formula https://en.wikipedia.org/wiki/Great-circle_distance .
- *  Throws exception when one or several input values are not within reasonable bounds.
- *  Latitude must be in [-90, 90], longitude must be [-180, 180].
- *  Original code of this implementation of this function is here https://github.com/sphinxsearch/sphinx/blob/409f2c2b5b2ff70b04e38f92b6b1a890326bad65/src/sphinxexpr.cpp#L3825.
- *  Andrey Aksenov, the author of original code, permitted to use this code in ClickHouse under the Apache 2.0 license.
- *  Presentation about this code from Highload++ Siberia 2019 is here https://github.com/ClickHouse/ClickHouse/files/3324740/1_._._GEODIST_._.pdf
- *  The main idea of this implementation is optimisations based on Taylor series, trigonometric identity and calculated constants once for cosine, arcsine(sqrt) and look up table.
- */
+/** Calculates the distance between two geographical locations.
+  * There are two variants:
+  * greatCircleDistance: calculates the distance on a sphere: https://en.wikipedia.org/wiki/Great-circle_distance
+  * geoDistance: calculates the distance on WGS-84 ellipsoid.
+  *
+  * The function calculates distance in meters between two points on Earth specified by longitude and latitude in degrees.
+  *
+  * Latitude must be in [-90, 90], longitude must be [-180, 180].
+  *
+  * Original code of this implementation of this function is here:
+  * https://github.com/sphinxsearch/sphinx/blob/409f2c2b5b2ff70b04e38f92b6b1a890326bad65/src/sphinxexpr.cpp#L3825.
+  * Andrey Aksenov, the author of original code, permitted to use this code in ClickHouse under the Apache 2.0 license.
+  * Presentation about this code from Highload++ Siberia 2019 is here https://github.com/ClickHouse/ClickHouse/files/3324740/1_._._GEODIST_._.pdf
+  * The main idea of this implementation is optimisations based on Taylor series, trigonometric identity
+  *  and calculated constants once for cosine, arcsine(sqrt) and look up table.
+  */
 
 namespace
 {
 
 constexpr double PI = 3.14159265358979323846;
-constexpr float TO_RADF = static_cast<float>(PI / 180.0);
-constexpr float TO_RADF2 = static_cast<float>(PI / 360.0);
+constexpr float RAD_IN_DEG = static_cast<float>(PI / 180.0);
+constexpr float RAD_IN_DEG_HALF = static_cast<float>(PI / 360.0);
+
+constexpr size_t COS_LUT_SIZE = 1024; // maxerr 0.00063%
+constexpr size_t ASIN_SQRT_LUT_SIZE = 512;
+constexpr size_t METRIC_LUT_SIZE = 1024;
+
+/// Earth mean diameter in meters, https://en.wikipedia.org/wiki/Earth
+constexpr float EARTH_DIAMETER = 2 * 6371000;
```

Comment:
Please, can you keep WGS-84 Earth radius from old implementation?
https://github.com/ClickHouse/ClickHouse/blob/v19.17.5.18-stable/dbms/src/Functions/greatCircleDistance.cpp#L14

Reply:
I have no idea what was that radius. There was no comments about it. Does it conform to some standard or to astronomical data?

```
Physical characteristics

mean radius  6371.0 km
equatorial     6378.1 km
polar radius  6356.8 km
```

Reply:
I found that it is "quadratic mean earth radius".
I need to spent a few minutes to figure out, what does it mean and by what measure the mean is taken...

Reply:
OK,  was wrong neither current, nor new radius are not correspondent to WGS 84. Probably it is better to use the same WGS 84 authalic radius (the radius of the sphere with the same area as WGS 84 ellipsoid) as H3 is used, because H3 is already a part of ClickHouse.
https://github.com/uber/h3/blob/master/src/h3lib/include/constants.h#L58
 
BTW, it is better to have angular distance function because it doesn't depend on spherical Earth radius definition.

Reply:
See table 3.3 here: 
https://earth-info.nga.mil/GandG/publications/tr8350.2/wgs84fin.pdf

Reply:
I have found it very unclear.
https://github.com/ClickHouse/ClickHouse/pull/8086/commits/1e3b9af66f567f0e6a266fd24bbc5ed403715003

But I can change to the radius from H3 :)

Reply:
Yes, we can add one more function. The name `greatCircleAngle` sounds Ok.